File: Sources/Features/Session/UseCases/ActivateFirstProjectUseCase.swift
// MARK: - ActivateFirstProjectUseCase
// Activates the first available project when no session is saved.
// macOS 14+, Swift 5.10

import Foundation
import OSLog

/// Activates the first available project and opens a terminal session for it.
///
/// Called when no previous session exists (first launch) or as a fallback
/// when the restored snapshot did not produce an active project selection.
/// The use case is a no-op when `activeProjectId` is already set or when
/// the project list is empty.
///
/// - Note: Must run on `@MainActor` because both `ProjectManaging` and
///   `TerminalSessionManaging` are `@MainActor`-isolated.
@MainActor
struct ActivateFirstProjectUseCase {

    // MARK: - Dependencies

    private let projectManager: any ProjectManaging
    private let terminalManager: any TerminalSessionManaging

    // MARK: - Init

    /// Creates a use case wired to the provided service implementations.
    ///
    /// - Parameters:
    ///   - projectManager: Checked for an existing active selection and mutated to set one.
    ///   - terminalManager: Receives the initial terminal session for the activated project.
    init(projectManager: any ProjectManaging, terminalManager: any TerminalSessionManaging) {
        self.projectManager = projectManager
        self.terminalManager = terminalManager
    }

    // MARK: - Execute

    /// Activate the first project if no project is currently active.
    ///
    /// Terminal session creation errors are silently discarded — the project
    /// is still activated even when the PTY cannot be spawned immediately.
    func execute() {
        guard projectManager.activeProjectId == nil,
              let first = projectManager.projects.first else { return }
        projectManager.activeProjectId = first.id
        try? terminalManager.createSession(
            for: first.id,
            shell: first.shellPath,
            workingDirectory: first.path
        )
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/usecase/ActivateFirstProjectUseCase.kt
package studio.vibe.shared.usecase

import studio.vibe.shared.contract.ProjectManaging
import studio.vibe.shared.contract.TerminalSessionManaging
import kotlin.uuid.ExperimentalUuidApi

/**
 * Activates the first available project and opens a terminal session for it.
 *
 * Called when no previous session exists (first launch) or as a fallback when the
 * restored snapshot did not produce an active project selection. The use case is a
 * no-op when [ProjectManaging.activeProjectId] is already set or when the project
 * list is empty.
 */
@OptIn(ExperimentalUuidApi::class)
class ActivateFirstProjectUseCase(
    private val projectManager: ProjectManaging,
    private val terminalManager: TerminalSessionManaging,
) {
    /**
     * Activate the first project if no project is currently active.
     *
     * Terminal session creation errors are silently discarded — the project is still
     * activated even when the PTY cannot be spawned immediately.
     */
    fun execute() {
        if (projectManager.activeProjectId.value != null) return
        val first = projectManager.projects.value.firstOrNull() ?: return

        projectManager.setActiveProjectId(first.id)

        // Guard: RestoreSessionUseCase may have already created a terminal for this
        // project. This happens when the snapshot's activeProjectId was a FreeTab
        // UUID (not persisted across restarts), leaving activeProjectId unset while
        // still restoring sessions for all real projects. Creating another session
        // here would produce an unwanted split with two duplicate terminals.
        if (terminalManager.sessions(first.id).isNotEmpty()) return

        try {
            terminalManager.createSession(
                projectId = first.id,
                shell = first.shellPath,
                workingDirectory = first.path,
            )
        } catch (e: Exception) {
            // Silently discard — project is still activated.
            println("ActivateFirstProjectUseCase: failed to create terminal session: ${e.message}")
        }
    }
}
